test/Passes/pass-rewrite-batch-matmul-to-matmul.mlir

```diff
@@ -74,16 +70,18 @@ func.func @batch_matmul_rewrite(%arg0: tensor<?x?x?xf32>,
 // CHECK-DAG: %[[C0:.+]] = arith.constant 0 : index
 // CHECK: %[[EMPTY:.+]] = tensor.empty(%[[ARG4]], %[[ARG2]], %[[ARG3]]) : tensor<?x?x?xf32>
 // CHECK: %[[DIM:.+]] = tensor.dim %[[ARG0]], %[[C0]] : tensor<?x?x?xf32>
-// CHECK: %[[DIM0:.+]] = tensor.dim %[[ARG0]], %[[C1]] : tensor<?x?x?xf32>
-// CHECK: %[[DIM1:.+]] = tensor.dim %[[ARG0]], %[[C2]] : tensor<?x?x?xf32>
-// CHECK: %[[DIM2:.+]] = tensor.dim %[[ARG1]], %[[C2]] : tensor<?x?x?xf32>
+// CHECK: %[[DIM0:.+]] = tensor.dim %[[ARG0]], %[[C0]] : tensor<?x?x?xf32>
+// CHECK: %[[DIM1:.+]] = tensor.dim %[[ARG0]], %[[C1]] : tensor<?x?x?xf32>
+// CHECK: %[[DIM2:.+]] = tensor.dim %[[ARG0]], %[[C2]] : tensor<?x?x?xf32>
+// CHECK: %[[DIM3:.+]] = tensor.dim %[[ARG1]], %[[C2]] : tensor<?x?x?xf32>
 // CHECK: %{{.+}} = scf.forall (%[[ARG5:.+]]) in (%[[DIM]]) 
 // CHECK-SAME:  shared_outs(%[[ARG6:.+]] = %[[EMPTY]]) -> (tensor<?x?x?xf32>) {
-// CHECK: %[[SLICE:.+]] = tensor.extract_slice %[[ARG0]][%[[ARG5]], 0, 0] [1, %[[DIM0]], %[[DIM1]]] [1, 1, 1] 
-// CHECK-SAME:  : tensor<?x?x?xf32> to tensor<?x?xf32>
-// CHECK: %[[SLICE1:.+]] = tensor.extract_slice %[[ARG1]][%[[ARG5]], 0, 0] [1, %[[DIM1]], %[[DIM2]]] [1, 1, 1] 
-// CHECK-SAME:  : tensor<?x?x?xf32> to tensor<?x?xf32>
-// CHECK: %[[SLICE2:.+]] = tensor.extract_slice %[[ARG6]][%[[ARG5]], 0, 0] [1, %[[DIM0]], %[[DIM2]]] [1, 1, 1] 
-// CHECK-SAME:  : tensor<?x?x?xf32> to tensor<?x?xf32>
-// CHECK: %{{.+}} = linalg.matmul ins(%[[SLICE]], %[[SLICE1]] : tensor<?x?xf32>, tensor<?x?xf32>) 
-// CHECK-SAME:  outs(%[[SLICE2]] : tensor<?x?xf32>) -> tensor<?x?xf32>
+// CHECK: %[[MIN:.+]] = affine.min #[[MAP]](%[[ARG5]])[%[[DIM0]]]
+// CHECK: %[[SLICE:.+]] = tensor.extract_slice %[[ARG0]][%[[ARG5]], 0, 0] [%[[MIN]], %[[DIM1]], %[[DIM2]]] [1, 1, 1] 
+// CHECK-SAME:  : tensor<?x?x?xf32> to tensor<?x?x?xf32>
+// CHECK: %[[SLICE1:.+]] = tensor.extract_slice %[[ARG1]][%[[ARG5]], 0, 0] [%[[MIN]], %[[DIM2]], %[[DIM3]]] [1, 1, 1] 
+// CHECK-SAME:  : tensor<?x?x?xf32> to tensor<?x?x?xf32>
+// CHECK: %[[SLICE2:.+]] = tensor.extract_slice %[[ARG6]][%[[ARG5]], 0, 0] [%[[MIN]], %[[DIM1]], %[[DIM3]]] [1, 1, 1] 
+// CHECK-SAME:  : tensor<?x?x?xf32> to tensor<?x?x?xf32>
+// CHECK: %{{.+}} = linalg.batch_matmul ins(%[[SLICE]], %[[SLICE1]] : tensor<?x?x?xf32>, tensor<?x?x?xf32>) 
```

Comment:
Is this doing the original `batch_matmul` N times for a `tensor<Nx?x?>`, in parallel?

Reply:
Sorry got stuck chasing a bug. It executes a batch_matmul with batch = 1 for N times. It seems that a canonicalization is missing to remove the affine.min and rank-reduce the tensor.

Reply:
Could it be because you pass as a separate argument and the compiler cannot prove that `%batch == dim(%arg0, 0)`? 

This kernel looks fragile, I would not write it that way in the first place.

Perhaps if you take the dim directly form the inputs, it can prove that the `min` is redundant and collapse the `matmul`.

Reply:
The affine min should not be there `affine_map<(d0)[s0] -> (-d0 + s0, 1)>` is always 1, as d0 goes from 0 to DIM0 - 1 and you subtract DIM0. Upstream tiling using scf.forall may miss some canonicalization for affine min, but I don't know which one :)

Reply:
It seems `canonicalize` does work on min/max:
https://github.com/llvm/llvm-project/blob/main/mlir/test/Dialect/Affine/canonicalize.mlir#L1104

Have you tried adding it as a following pass in the RUN command line?

Reply:
yes, canonicalize in this case does not help.

Reply:
Can we make canonicalise work on this upstream then update this? I'm ok with a TODO for now if we create an issue and work on it soon.